### PR TITLE
[Storage] Error Handle Attempts to Mount Non-sky Managed Bucket without Source

### LINKED ISSUE #2779: [Storage] sky start failing to re-mount on externally created Storage
User reported to @Michaelvll :

> Got this error while restarting an instance:
> `sky.exceptions.StorageExternalDeletionError: The bucket, 'bucket-name', could not be mounted on cluster 'cluster-name'. Please verify that the bucket exists. The cluster started successfully without mounting the bucket.`

User also confirmed that the storage exists in the console. Did some investigation and found that this is an edge case for cloud storages that are not sky managed (created externally). 

repro:
1. `sky launch` with a task yaml specifying `MOUNT` mode on externally created cloud storage(create storage manually from cloud service provider console)
2. `sky stop` the cluster
3. `sky start` the cluster

This edge case should be fixed so the externally created storages can be re-mounted when `sky start` is ran.

Update: This edge case is rather rooting from an undesired behavior of current `Storage` class implementation. To raise the above error by attempting to re-mount an externally created bucket with `sky start`, user specified a `name` of the externally created bucket under `file_mount`'s section when first launched the instance. However, we only want to allow mounting an externally created bucket by user specifying the bucket's url to the `source` field. The fix to this issue should be to raise an error when a `name` of externally created bucket is specified with `MOUNT` mode.

### PR DESCRIPTION
**Pull request by landscapepainter**
This closes #2779, #2950

`Storage` is designed in a way to not allow externally created cloud storages to be mounted without specifying the `source` field. The only way to mount non-sky managed bucket is to specify the bucket's URI as the `source` field. This fix raises `StorageSpecError` when a non-sky managed bucket is attempted to be mounted just with `name` field and without specifying the `source` field.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] All smoke tests: `pytest tests/test_smoke.py::TestStorageWithCredentials` 
- [x] Manually test: Create storage externally and `MOUNT` with `name` specified and `source` unspecified. Should raise error
- [x] Manually test: Create storage externally and `MOUNT` with `source` specified and `name` unspecified. Should go through
- [x] `pytest tests/test_smoke.py::test_gcp_storage_mounts_with_stop --gcp`
- [x] `pytest tests/test_smoke.py::test_aws_storage_mounts_with_stop --aws`

### CODE REVIEW

**Review by Michaelvll — COMMENTED** [review]
Thanks for submitting this PR @landscapepainter! Can we list out all the syntax allowed for a storage object in the docstr of the module? It has been quite complicated for disallowing some of the combination for the fields.

I think one decision we made before was that we allow the following syntax:
The following should be allowed for both sky managed and external buckets.
```yaml
/destination:
  source: s3://xxx
```
```yaml
/destination:
  name: store-name
  stores: s3
```

The only one we should raise error for is the following:
```yaml
/destination:
  name: store-name
  # without stores
  mode: COPY
```

cc'ing @romilbhardwaj @concretevitamin for confirmation.

File: sky/data/storage.py [hunk 1]
```diff
@@ -1248,6 +1254,26 @@ def _get_bucket(self) -> Tuple[StorageHandle, bool]:
...
+                # externally created buckets, users must provide the
+                # bucket's URL as 'source'.
+                if handle is None:
+                    with ux_utils.print_exception_no_traceback():
+                        raise exceptions.StorageSpecError(
+                            'Attempted to mount a non-sky managed bucket '
+                            f'{self.name!r} from S3 without specifying a '
+                            'storage source. To mount an externally created '
+                            'bucket, please specify the bucket URL as '
+                            'the storage source.')
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/data/storage.py]: Refactored at [a602af3](https://github.com/skypilot-org/skypilot/pull/2804/commits/a602af39e6c45f49d90371c11fc24e8e787afaeb)

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter - left some minor comments. Please also make sure all spot smoke tests pass since the spot controller relies on storage quite a bit for file_mounts translation.

Regarding allowing this (where `store-name` is an externally created bucket):
```
/destination:
  name: store-name
  stores: s3
```

@Michaelvll @concretevitamin - I thought about this and concluded that this should not be allowed. Allowing this would make intent ambiguous - it's unclear if the user wants to:

a) create `store-name` from scratch or 
b) mount `s3://store-name`, where `store-name` is an existing bucket. 

This ambiguity is dangerous when the user:

a) Wants to create a bucket but it already exists -> SkyPilot will silently mount an existing bucket, which may potentially be a public bucket. Instead it should raise an error saying it already exists (which it does today)
b) Wants to use an existing bucket but has a typo (e.g., `name: storename` instead of `name: store-name`) -> SkyPilot will silently create a new empty bucket for the user and the user's task will error if it tries to read some files it expects to be there.

As you can see, the root cause is stemming from our attempt at having a "get or create" interface through `file_mounts`, which introduces this ambiguity unless explicitly handled by our code, such as in this PR. 

Our early Sky Storage proposals took inspiration from Kubernetes and used two separate fields in the YAML - one to declare any new Storages, and other to attach storage to task/cluster. This gets rid of ambiguity, but was harder to use. Thus this is the tradeoff we chose, and we minimize ambiguity here by clarifying intent when required.

File: sky/data/storage.py [hunk 2]
```diff
@@ -321,6 +321,28 @@ def __deepcopy__(self, memo):
         # original Store object is returned
         return self
 
+    def _validate_existing_bucket(self):
+        """Validates the storage fields for existing buckets."""
+        # Check if 'source' is None, indicating Storage is in MOUNT mode.
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/data/storage.py:326]: @romilbhardwaj This is one of the cases we don't allow the users. We don't allow users to only provide the `name` of the bucket without `source`.

File: sky/data/storage.py [hunk 3]
```diff
@@ -321,6 +321,28 @@ def __deepcopy__(self, memo):
...
+            # externally created buckets, users must provide the
+            # bucket's URL as 'source'.
+            if handle is None:
+                with ux_utils.print_exception_no_traceback():
+                    raise exceptions.StorageSpecError(
+                        'Attempted to mount a non-sky managed bucket '
+                        f'{self.name!r} from S3 without specifying a '
+                        'storage source. To mount an externally created '
+                        'bucket, please specify the bucket URL as '
+                        'the storage source.')
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/data/storage.py]: resolved in [ee2a8bd](https://github.com/skypilot-org/skypilot/pull/2804/commits/ee2a8bde9ca0b9fa517f592bb8cc6d6fe1e6b917)

**Review by landscapepainter — COMMENTED** [review]
This is ready for another look! Regards to listing allowed syntax for storage, I'm wondering if it'd be nice to have it in our readthedoc instead of the docstr of `Storage` class so the users can easily access to use them as a reference as well? @Michaelvll @romilbhardwaj @concretevitamin

**Comment by landscapepainter** [comment]
From offline discussion with @romilbhardwaj, it was discussed the following tasks should be done:
1. Check if the following works for sky managed already existing storage:
```
file_mounts: 
  name: myexistingstorage
  mode: COPY
```
2.  Fix the following comment since the `'source'` being `None` does not indicate `MOUNT` mode. It can also be the case with `COPY` mode for already existing sky managed storage.
`# Check if 'source' is None, indicating Storage is in MOUNT mode.`
3. Write smoke test to make sure this case works if it does not exist already.

**Comment by landscapepainter** [comment]
Writing this for future reference on why the handling for this edge case cannot be done in `_validate_storage_spec`:

We allow to create new `MOUNT` mode storages without `source` specified and with `name` only, which is when user wants to use the bucket as a scratch space. On the other hand, we don't allow to `MOUNT` non-sky managed Storages with `name` only and  require `source` to do so. In either case, `global_user_state` does not contain the hanlde of the `Storage`. And this is the key reason why we cannot validate this edge case from `_validate_storage_spec`.

To make sure we don't block users from creating new `MOUNT` mode storage with only specifying `name` and without `source`, we need to make sure not to merely block this early from `_validate_storage_spec` and  check if the storage exists first from `_get_bucket`(thorugh methods such as `self.client.head_bucket()`) and then check if it's sky managed through existance of the `handle` in `global_user_state`. If it's not sky managed, this is gauranteed to be a storage created externally and not sky managed, which is the case we should block the user's `MOUNT`ing without `source` and with just `name`.

**Comment by landscapepainter** [comment]
@romilbhardwaj This is ready for another look!

**Comment by Michaelvll** [comment]
Any update for this?

**Comment by landscapepainter** [comment]
@Michaelvll This PR is waiting for the Storage page from readthedoc to be reformatted. @romilbhardwaj is working on it.

**Comment by romilbhardwaj** [comment]
Blocked on #3123

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter! This works nicely - tested with smoke tests and manual tests. Will approve and merge once #3162 is merged.
